core/core.go

```diff
@@ -731,8 +739,35 @@ func constructPeerHost(ctx context.Context, id peer.ID, ps pstore.Peerstore, bwr
 		hostOpts = append(hostOpts, p2pbhost.NATPortMap)
 	}
 
+	if !opts.DisableRelay {
+		filterRelayAddr := func(addrs []ma.Multiaddr) []ma.Multiaddr {
+			var raddrs []ma.Multiaddr
+			for _, addr := range addrs {
+				_, err := addr.ValueForProtocol(circuit.P_CIRCUIT)
+				if err == nil {
+					continue
+				}
+				raddrs = append(raddrs, addr)
+			}
+			return raddrs
+		}
+		hostOpts = append(hostOpts, p2pbhost.AddrsFactory(filterRelayAddr))
```

Comment:
Added the scaffolding for address factory composition in https://github.com/ipfs/go-ipfs/pull/4091/commits/231f8e46b2ac26a2b344c958bbe2765c0ea41961:
- added an AddrsFactory option to ConstructPeerHostOpts
- composed the specified address factory with the relay filter when present